iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,581 @@
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::insert_after(const_iterator citer,
+                                                                                     T&& ur_data) noexcept
+{
+    return emplace_after(citer, std::forward<T>(ur_data));
+}
+
+
+template <typename T, int64_t CAPACITY>
+void forward_list<T, CAPACITY>::clear() noexcept
+{
+    for (auto iter = before_begin(); begin() != end();)
+    {
+        erase_after(iter);
+        iter = before_begin();
+    }
+}
+
+/*************************/
+// private member functions
+/*************************/
+
+/*************************/
+// iterator
+template <typename T, int64_t CAPACITY>
+forward_list<T, CAPACITY>::iterator::iterator(forward_list* parent, size_type idx) noexcept
+    : m_list(parent)
+    , m_iterListNodeIdx(idx)
+{
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator& forward_list<T, CAPACITY>::iterator::operator++() noexcept
+{
+    // get node ref from List
+    if (m_list->isValidIteratorIndex(m_iterListNodeIdx))
+    {
+        ListNodePointer const nodePtr = m_list->getNodePtrFromIdx(m_iterListNodeIdx);
+        m_iterListNodeIdx = nodePtr->nextIdx;
+        return *this;
+    }
+    else
+    {
+        // do not increase beyond end() position (invalid operation), but return as-is
+        errorMessage(__PRETTY_FUNCTION__, "invalid iterator or end()");
+        return *this;
+    }
+}
+
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::iterator::operator==(const iterator rhs_iter) const noexcept
+{
+    // break recursive call for iterator::operator== by converting at least one parameter to const_iterator
+    return (rhs_iter == const_iterator{*this});
+}
+template <typename T, int64_t CAPACITY>
+bool forward_list<T, CAPACITY>::iterator::operator==(const const_iterator rhs_citer) const noexcept
+{
+    return (rhs_citer == *this);
+}
+
+template <typename T, int64_t CAPACITY>
+T& forward_list<T, CAPACITY>::iterator::operator*() const noexcept
+{
+    if (!m_list->isValidElementIndex(m_iterListNodeIdx))
+    {
+        // terminate e.g. when trying to read e.g. from end()
+        errorMessage(__PRETTY_FUNCTION__, " malformed iterator");
+        std::terminate();
+    }
+    const ListNodePointer nodePtr = m_list->getNodePtrFromIdx(m_iterListNodeIdx);
+    return nodePtr->data;
+} // operator*
+
+
+template <typename T, int64_t CAPACITY>
+T* forward_list<T, CAPACITY>::iterator::operator->() const noexcept
+{
+    return &operator*();
+}
+// iterator
+/*************************/
+// const_iterator
+
+template <typename T, int64_t CAPACITY>
+forward_list<T, CAPACITY>::const_iterator::const_iterator(const forward_list* parent, size_type idx) noexcept
+    : m_cList(parent)
+    , m_iterListNodeIdx(idx)
+{
+}
+
+template <typename T, int64_t CAPACITY>
+forward_list<T, CAPACITY>::const_iterator::const_iterator(const iterator iter) noexcept
```

Comment:
I think we want ```const_iterator(const iterator& iter) noexcept;``` here. Iterators are usually small, but not necessarily value only objects. This allows us to convert iterators to const_iterators which is useful for comparison as well.